### PR TITLE
Fix/refactor url service

### PR DESCRIPTION
**Pull request by MazanYan**
* Moved URL and authorization exceptions to separate categories
* Removed sample UsersException and UrlsException that are not used
* Moved short link to separate DTO
* Moved list of short links (for `GET /urls` controller) to separate DTO
* Store URL in the short link as a string: for easier validation and serialization (temporarily using Gson - must switch to Jackson in future)
* Renamed fiends in ShortLink DTO (to make serialized JSON correspond to the project specification)
* Own URL validation in `safelyCreateUrl` method of ShortLinkService

Still need a clean way to delete links and users after testing

### CODE REVIEW

File: src/main/java/edu/kpi/testcourse/logic/ShortLinkServiceImpl.java [hunk 1]
```diff
@@ -111,11 +108,10 @@ public String generateAlias() {
    * @param destination URL that should be validated
    * @return URL if it can be created
    */
-  public Optional<URL> safelyCreateUrl(String destination) {
-    try {
-      URL url = new URL(destination);
-      return Optional.of(url);
-    } catch (MalformedURLException e) {
+  public Optional<String> safelyCreateUrl(String destination) {
```
Comment (RingaVirda) [inline src/main/java/edu/kpi/testcourse/logic/ShortLinkServiceImpl.java:111]: I don't understand the point of this method. Is it just validation? If so please think about renaming it to for ex "validateUrl" which would return optional.empty if invalid

File: src/main/java/edu/kpi/testcourse/logic/ShortLinkServiceImpl.java [hunk 2]
```diff
@@ -111,11 +108,10 @@ public String generateAlias() {
...
+    if (destination.startsWith("http://") || destination.startsWith("https://")) {
```
Comment (RingaVirda) [inline src/main/java/edu/kpi/testcourse/logic/ShortLinkServiceImpl.java:112]: If it is validation maybe you should use something like that:
        try { 
            new URL(url).toURI(); 
            return true; 
        } 
        catch (Exception e) { 
            return false; 
        } 
(Took this code from somewhere) Cause if we'll stick with your version first of all something like "http://afnaonfa" will be valid and also it'll be restricted to this two protocols

File: src/main/java/edu/kpi/testcourse/rest/UrlController.java [hunk 3]
```diff
@@ -36,18 +35,6 @@
...
-  public record UserUrl(String url, String alias) {}
-
-  /**
-   * Standard response body for GET /urls.
-   */
-  @Introspected
-  public record UserUrls(ArrayList<ShortLinkMock> urls) {}
-
   /**
    * Authorized user cen shorten an URL on this route.
```
Comment (RingaVirda) [inline src/main/java/edu/kpi/testcourse/rest/UrlController.java:39]:
```suggestion
   * Authorized user can shorten a URL on this route.
```

File: src/test/java/edu/kpi/testcourse/rest/RedirectLinkTest.java [hunk 4]
```diff
@@ -41,21 +38,40 @@ public static void setupServer() {
...
-    String withAliasBody = client.toBlocking().retrieve(
-      HttpRequest.POST("/urls/shorten", new UserUrl(
-        "https://darcs.realworldhaskell.org/static/00book.pdf",
-        "haskell"
-      )).header("token", TEST_VALID_TOKEN)
-    );
-
-    randomAlias = getShortenedUrlFromResponseBody(noAliasBody);
-    customAlias = getShortenedUrlFromResponseBody(withAliasBody);
+    assertDoesNotThrow(() -> {
```
Comment (RingaVirda) [inline src/test/java/edu/kpi/testcourse/rest/RedirectLinkTest.java:41]: AssertDoesNotThrow for the win!

File: src/test/java/edu/kpi/testcourse/rest/ShortenLinkTest.java [hunk 5]
```diff
@@ -111,15 +155,33 @@ public void shouldNotSaveInvalidUrl() {
...
 
     assertThat(body.contains("\"shortened_url\":\"http://localhost:8080/r/alias\"")).isEqualTo(true);
+
+    // clear all links in 'links' directory
+    // temporary solution before we configure our data storage properly
+    assertDoesNotThrow(() -> {
+      Path linksPath = Paths.get("data/links/");
+
+      try (Stream<Path> walk = Files.walk(linksPath)) {
+        walk.sorted(Comparator.reverseOrder())
```
Comment (RingaVirda) [inline src/test/java/edu/kpi/testcourse/rest/ShortenLinkTest.java:181]: Check if it does not delete the data/links folder

File: src/test/java/edu/kpi/testcourse/rest/ShortenLinkTest.java [hunk 6]
```diff
@@ -1,26 +1,36 @@
...
 import com.google.gson.Gson;
-import edu.kpi.testcourse.rest.UrlController.UserUrl;
+import edu.kpi.testcourse.dto.ShortLink;
 import io.micronaut.context.ApplicationContext;
 import io.micronaut.http.HttpRequest;
 import io.micronaut.http.client.HttpClient;
 import io.micronaut.http.client.exceptions.HttpClientResponseException;
 import io.micronaut.runtime.server.EmbeddedServer;
+import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
+import java.io.File;
```
Comment (RingaVirda) [inline src/test/java/edu/kpi/testcourse/rest/ShortenLinkTest.java:11]: Why do you need these while testing ShortenLink logic?

File: src/test/java/edu/kpi/testcourse/rest/ShortenLinkTest.java [hunk 7]
```diff
@@ -1,26 +1,36 @@
...
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
+@MicronautTest
```
Comment (RingaVirda) [inline src/test/java/edu/kpi/testcourse/rest/ShortenLinkTest.java:26]: Make a MicronautTest and never inject something

File: src/main/java/edu/kpi/testcourse/dto/ShortLink.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
+package edu.kpi.testcourse.dto;
+
+import io.micronaut.core.annotation.Introspected;
+import java.net.URL;
+
+/**
+ * Mock data for short link record in the data storage.
```
Comment (RingaVirda) [inline src/main/java/edu/kpi/testcourse/dto/ShortLink.java:7]:
```suggestion
 * Entity for short link record in the data storage.
```

**Review by RingaVirda — COMMENTED** [review]
Good refactor
Left some questions in comments

File: src/test/java/edu/kpi/testcourse/rest/ShortenLinkTest.java (hunk 5 above)
Comment (MazanYan) [reply src/test/java/edu/kpi/testcourse/rest/ShortenLinkTest.java:181]: Yes, it does not delete the folder, only files included

File: src/main/java/edu/kpi/testcourse/logic/ShortLinkServiceImpl.java (hunk 1 above)
Comment (MazanYan) [reply src/main/java/edu/kpi/testcourse/logic/ShortLinkServiceImpl.java:111]: Thanks for a good point! Forgot to rename it when I switched from `java.net.URL` to `String` to store URLs.

File: src/main/java/edu/kpi/testcourse/logic/ShortLinkServiceImpl.java (hunk 2 above)
Comment (MazanYan) [reply src/main/java/edu/kpi/testcourse/logic/ShortLinkServiceImpl.java:112]: I agree that `java.net.URL` is better for validation than validation via regular expressions or other ways. Will update the method then.

File: src/test/java/edu/kpi/testcourse/rest/RedirectLinkTest.java (hunk 4 above)
Comment (MazanYan) [reply src/test/java/edu/kpi/testcourse/rest/RedirectLinkTest.java:41]: Maybe I forgot to remove this assertion after I switched from `URL` to `String` to store URLs.

**Review by RingaVirda — APPROVED** [review]
Thx for accepting suggested changes, approve from me
Maybe someone else should review this before merge

**Pull request merged**